src/nvim/edit.c

```diff
@@ -275,7 +275,10 @@ static void insert_enter(InsertState *s)
 
     set_vim_var_string(VV_INSERTMODE, (char *) s->ptr, 1);
     set_vim_var_string(VV_CHAR, NULL, -1);
+
     apply_autocmds(EVENT_INSERTENTER, NULL, NULL, false, curbuf);
+    apply_autocmds(EVENT_MODEENTER, (char_u *)"insert", NULL, false, curbuf);
+    apply_autocmds(EVENT_MODELEAVE, (char_u *)"norm", NULL, false, curbuf);
```

Comment:
I don't think we want pairs like this, modes in nvim is more like a stack machine than a finite state machine. Like, when pressing `:` in normal mode, then cmdline mode is entered. Similarly, when `<c-f>` is pressed in cmdline mode, again normal mode is entered. But when `<esc>` is pressed in cmdline mode, the mode is instead left (i e popped of the stack). If we want to publish both old and new mode, use `v:event` is better.

Reply:
Hmm, I didn't had much idea of it, OK sure, I understand it, I will edit the code and will soon make a PR again.
Till now, I again make this PR as WIP